compiler/rustc_typeck/src/coherence/inherent_impls.rs

```diff
@@ -387,7 +412,34 @@ impl InherentCollect<'tcx> {
                     lang,
                     ty
                 )
-                .span_help(span, "consider using a trait to implement these methods")
+                .span_help(
+                    span,
+                    &format!("consider using a trait{}", {
+                        if assoc_items.len() == 0 {
+                            String::new()
+                        } else {
+                            let plural = assoc_items.len() > 1;
+                            format!(
+                                " to implement {} {}{}",
+                                if plural { "these" } else { "this" },
+                                {
+                                    let item_types = assoc_items.iter().map(|x| x.kind);
+                                    if item_types.clone().all(|x| x == hir::AssocItemKind::Const) {
+                                        "constant"
+                                    } else if item_types
+                                        .clone()
+                                        .all(|x| matches! {x, hir::AssocItemKind::Fn{ .. } })
+                                    {
+                                        "method"
+                                    } else {
+                                        "associated item"
+                                    }
+                                },
+                                if plural { "s" } else { "" }
+                            )
+                        }
```

Comment:
I think ideally you would extract these computations out to before the `struct_span_err!` call. Then you would just reference local variables in the `format!` call. The way it is right now is quite hard to read, since it has a lot of nested computations inside nested `format!` calls.